src/gpu/kmeans/kmeans_init.cu

```diff
@@ -0,0 +1,439 @@
+/*!
+ * Copyright 2018 H2O.ai, Inc.
+ * License   Apache License Version 2.0 (see LICENSE for details)
+ */
+
+#include <thrust/device_vector.h>
+
+#include <cub/device/device_histogram.cuh>
+
+#include <random>
+#include <limits>
+#include <string>
+
+#include <stdio.h>
+
+#include "kmeans_init.cuh"
+
+#include "../matrix/KmMatrix/KmMatrix.hpp"
+#include "../matrix/KmMatrix/Arith.hpp"
+#include "../matrix/KmMatrix/blas.cuh"
+#include "../utils/utils.cuh"
+#include "../utils/GpuInfo.cuh"
+
+namespace h2o4gpu {
+namespace kMeans {
+
+using namespace Matrix;
+
+namespace kernel {
+// X^2 + Y^2, here only calculates the + operation.
+template <typename T>
+__global__ void construct_distance_pairs_kernel(
+    kParam<T> _distance_pairs,
+    kParam<T> _data_dots, kParam<T> _centroids_dots) {
+
+  size_t idx = global_thread_idx();  // indexing data
+  size_t idy = global_thread_idy();  // indexing centroids
+
+  // FIXME: Is using shared memory necessary?
+
+  size_t stride_x = grid_stride_x ();
+  // strides only for data.
+  for (size_t i = idx; i < _data_dots.rows; i += stride_x) {
+    if (idy < _centroids_dots.rows ) {
+      // i + idy: x^2 + y^2 between i^th data (a.k.a x) and idy^th
+      // centroid (a.k.a y)
+      _distance_pairs.ptr[i*_centroids_dots.rows + idy] =
+          _data_dots.ptr[idx] + _centroids_dots.ptr[idy];
+    }
+  }
+}
+
+// See SelfArgMinOp
+template <typename T>
+__global__ void self_row_argmin_sequential(kParam<int> _res, kParam<T> _val) {
+
+  size_t idx = global_thread_idx();
+  if (idx < _val.rows) {
```

Comment:
Consider doing this: `if (idx >= _val.rows) return;`

I find it a bit cleaner but it's a matter of personal preference.